### DISCUSSION TITLE
How to bind Publishers correctly on Origin Server?

### DISCUSSION THREAD

**Discussion by FRahmann** [OvenMediaLabs/OvenMediaEngine#1671] [Help] [answered]
im trying to have multiple docker container containing origin configs to have multiple servers in my testbench. im getting this error:

> [2024-07-27 23:53:44.384] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:73   | Server ID : 936df643-4f49-4fe2-aaf6-064b1aa8a99f
> [2024-07-27 23:53:44.507] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:91   | Resolved public IP address (XXX.XXX.XXX.XXX) from stun server (stun.ovenmediaengine.com:13478)
> [2024-07-27 23:53:44.507] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:109  | This host supports IPv4 only (reason: v6: Failed to bind(): [errno] Cannot assign requested address (99)[0])
> [2024-07-27 23:53:44.513] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:126  | Trying to create MediaRouter...
> [2024-07-27 23:53:44.513] I [OvenMediaEngine:1] MediaRouter | mediarouter.cpp:40   | Mediarouter has been started.
> [2024-07-27 23:53:44.513] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:129  | Trying to create WebRTC Publisher...
> [2024-07-27 23:53:44.513] W [OvenMediaEngine:1] WebRTC Publisher | webrtc_publisher.cpp:105  | WebRTC Publisher is disabled by configuration
> [2024-07-27 23:53:44.513] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:130  | Trying to create LLHLS Publisher...
> [2024-07-27 23:53:44.513] W [OvenMediaEngine:1] LLHLS Publisher | llhls_publisher.cpp:78   | LLHLS Publisher is disabled by configuration
> [2024-07-27 23:53:44.514] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:131  | Trying to create OVT Publisher...
> [2024-07-27 23:53:44.514] E [OvenMediaEngine:1] Socket | socket.cpp:409  | [#7] [0x7fc2e2eecc90] Could not bind to XXX.XXX.XXX.XXX:9000 (-1, [errno] Cannot assign requested address (99))
> [2024-07-27 23:53:44.615] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:200  | Trying to release MediaRouter...
> [2024-07-27 23:53:44.615] I [OvenMediaEngine:1] MediaRouter | mediarouter.cpp:49   | Mediarouter has been stopped.
> [2024-07-27 23:53:44.614] E [OvenMediaEngine:1] OVT | ovt_publisher.cpp:77   | Could not listen for OVTPublisher on XXX.XXX.XXX.XXX:9000
> [2024-07-27 23:53:44.615] E [OvenMediaEngine:1] OvenMediaEngine | main.cpp:131  | Failed to initialize OVT Publisher
> [2024-07-27 23:53:44.615] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:326  | Uninitializing TCP socket pool...
> [2024-07-27 23:53:44.616] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:328  | Uninitializing UDP socket pool...
> [2024-07-27 23:53:44.712] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:331  | OvenMediaEngine will be terminated

with the following config:

```
<?xml version="1.0" encoding="UTF-8"?>

<Server version="8">
        <Name>o1.uav-nord.de</Name>
        <Type>origin</Type>
        <IP>XXX.XXX.XXX.XXX</IP>
        <PrivacyProtection>true</PrivacyProtection>
        <StunServer>stun.ovenmediaengine.com:13478</StunServer>
        <Modules>
                <HTTP2>
                        <Enable>true</Enable>
                </HTTP2>
                <LLHLS>
                        <Enable>true</Enable>
                </LLHLS>
        </Modules>
        <Bind>
                <Managers>
                        <API>
                                <Port>8081</Port>
                                <TLSPort>8082</TLSPort>
                                <WorkerCount>4</WorkerCount>
                        </API>
                </Managers>

                <Providers>
                        <RTSPC>
                                <WorkerCount>2</WorkerCount>
                        </RTSPC>
                        <OVT>
                                <WorkerCount>2</WorkerCount>
                        </OVT>
                        <RTMP>
                                <Port>1935</Port>
                                <WorkerCount>2</WorkerCount>
                        </RTMP>
                        <SRT>
                                <Port>9999</Port>
                                <WorkerCount>2</WorkerCount>
                        </SRT>
                </Providers>
                <Publishers>
                        <OVT>
                                <Port>9000</Port>
                                <WorkerCount>2</WorkerCount>
                        </OVT>
                </Publishers>
        </Bind>
        <Managers>
                <Host>
                        <Names>
                                <Name>instance1.my-fancy-domain</Name>
                                <Name>XXX.XXX.XXX.XXX</Name>
                        </Names>
                        <TLS>
                                <CertPath>/path/to/my/certs</CertPath>
                                <KeyPath>/path/to/my/certs</KeyPath>
                                <ChainCertPath>/path/to/my/certs</ChainCertPath>
                        </TLS>
                </Host>
                <API>
                        <AccessToken>******************************************************************</AccessToken>
                </API>
        </Managers>
</Server>
```

As i said, there are currently 2 docker containers configured with the same ports but different ip addresses. I cannot understand why the service cannot bind IP:PORT combination.. the additional ips are pingable , etc.

King regards

**Comment by naanlizard** [comment]
The error appears to be that you're trying to bind an ipv6 address and your host doesn't support that. What's the output of ifconfig?

**Reply by FRahmann** [reply]
I have no ipv6 support at all on this machine.
I disabled it completly. is there any way to force docker or the container not to mention v6?

**Reply by naanlizard** [reply]
Is that your full server. XML? Are you sure there are no typos in your IP addresses?

Does it matter which container of the two you bring up first? As in, is it always the second one that throws an error, or is it always one of the config versions, whether it is brought up first or second?

Also please use GitHub gists or something like that in the future, inline logs are harder to read

**Reply by FRahmann** [reply]
Point for you...If i clicked save it would work... 
Now i completly disabled v6 via sysctl.conf in my system. But there is still this error i cannot understand why:

`[2024-07-27 23:53:44.514] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:131  | Trying to create OVT Publisher...
[2024-07-27 23:53:44.514] E [OvenMediaEngine:1] Socket | socket.cpp:409  | [#7] [0x7fc2e2eecc90] Could not bind to XXX.XXX.XXX.XXX:9000 (-1, [errno] Cannot assign requested address (99))
[2024-07-27 23:53:44.615] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:200  | Trying to release MediaRouter...
[2024-07-27 23:53:44.615] I [OvenMediaEngine:1] MediaRouter | mediarouter.cpp:49   | Mediarouter has been stopped.
[2024-07-27 23:53:44.614] E [OvenMediaEngine:1] OVT | ovt_publisher.cpp:77   | Could not listen for OVTPublisher on XXX.XXX.XXX.XXX:9000
[2024-07-27 23:53:44.615] E [OvenMediaEngine:1] OvenMediaEngine | main.cpp:131  | Failed to initialize OVT Publisher
[2024-07-27 23:53:44.615] I [OvenMediaEngine:1] OvenMediaEngine | main.cpp:326  | Uninitializing TCP socket pool...`

**Comment by naanlizard** [comment]
Please share both full configuration files and the docker run commands or
docker compose file

**Reply by FRahmann** [reply]
https://gist.github.com/FRahmann/0871f1c722c98d08c443ecd64133aaee

didnt know about gists before. sry ;)

**Comment by naanlizard** [comment, marked as answer]
If you're not using your environmental variables, why define them? Try
removing them and see if it helps.

Also confirm the settings in the server.xml for any IP listed, especially
if there are colons in them

**Reply by FRahmann** [reply]
I removed all port definitions and setted the network mode to host to directly assign the addresses. afterwards it worked so far regarding this problem. Now im struggeling with TLS but thats on another page ;) Thanks for your time and your patience!
